# Install BigDFT code or client in Google Colab, locally or into your google drive folder
This notebook will install the code in your google drive folder, which will be useful to excute the other tutorials of the school.

You should provide authorisation to access your google drive after the execution of the first cell.

The files of the school will go in the `bigdft-school` directory. Such folder can me moved to your local disk or deleted once you are done.

The entire installation procedure should take about one minute.

In the forthcoming lessons, some other packages will be also installed in this folder.

In [ ]:
!wget https://gitlab.com/luigigenovese/bigdft-school/-/raw/main/packaging/install.py &> /dev/null
import install
install.full_suite() # Kernel will restart after this, it is normal. Count about 3 minutes

The notebook should have restarted at present. You may want to remove the runtime warning at the bottom of the page.

Also, you may have seen a message about an inconsistent environment at the end of the installation procedure. This message is harmless, you can safely ignore it.

The install module will have to be imported again to set the notebook ready for execution.

To see if everything looks OK you can execute this cell to see if an energy value is output at the end:

In [1]:
import install
install.set_ready() #this resets the basic environment variables and puts us in the correct directory, (only needed for the full suite installation)

Executing: mkdir -p /content/bigdft-school



In [4]:
# Test that we can import everything
import futile
import BigDFT

# Test A Full Calculation with PyBigDFT
from BigDFT.Systems import System
from BigDFT.Fragments import Fragment
from BigDFT.Atoms import Atom

at = Atom({"He": [0, 0, 0]})
frag = Fragment([at])
sys = System({"FRA:0": frag})

from BigDFT.Calculators import SystemCalculator
code = SystemCalculator(skip=True) #this skip option would not run the code if the result is present.

from BigDFT.Inputfiles import Inputfile
inp = Inputfile()
inp.set_xc("PBE")
inp.set_hgrid(0.4)

log = code.run(sys=sys, input=inp)
print(log.energy)



Initialize a Calculator with OMP_NUM_THREADS=1 and command /content/bigdft-school/install/bin/bigdft
Creating the yaml input file "./input.yaml"
Executing command:  /content/bigdft-school/install/bin/bigdft -s Yes
-2.8893983558799454


Such energy value should be `-2.8893983558799454` (Hartree)

We can now present another possible installation of the code, which is performed via only the bigdft client. This installation is lighter as it only involves the python pakages employed by PyBigDFT for pre- and post- processing, and it will be enough for most of the tutorials of the school.


In [3]:
import install
install.client(locally=True)

Executing: mkdir -p /content

Executing: git clone --depth 1 https://gitlab.com/luigigenovese/bigdft-school
Error Occurred:  
 fatal: destination path 'bigdft-school' already exists and is not an empty directory.

Executing: mkdir -p /content/bigdft-school



# Verification step
Rerun the execution step now that the client is installed instead to see if the energy is the same even without the BigDFT executable.

Tips: restart the runtime without erasing the files, and run the installation command. (`import install` first, then `install.client(locally=True)`)

# Installation of the client on your google drive

Install the bigdft client on your google drive folder to continue with the school trainings.

To trigger this installation you have to authorize the usage of your google drive folder. This would prevent the installation to be performed evey time.

Tips: it is just enough to remove the ``locally=True`` option.


In [5]:
install.client()

Mounted at /content/drive
Executing: mkdir -p /content/drive/MyDrive

Executing: git clone --depth 1 https://gitlab.com/luigigenovese/bigdft-school
Error Occurred:  
 fatal: destination path 'bigdft-school' already exists and is not an empty directory.

Executing: mkdir -p /content/drive/MyDrive/bigdft-school



# Installation form
In the forthcoming notebooks you have to setup the bigdft installation each time.
If you performed your installtion on google drive, this will just remount your drive and set the correct python environment. For the other installation flavours, it would be necessary to perform the installation steps from scratch.
We include a form at the beginning of each notebook to suggest the preferred installation method for each session. When suitable, you can choose another one.

If you want to test the form, restart the runtime, choose your installation method, and push the "Play" button.

In [7]:
install = "client (Google drive)" #@param ["full_suite", "client (Google drive)", "client"]
install_var=install
!wget https://gitlab.com/luigigenovese/bigdft-school/-/raw/main/packaging/install.py &> /dev/null
args={'locally': True} if install == 'client' else {}
import install
getattr(install,install_var.split()[0])(**args)

Mounted at /content/drive
Executing: mkdir -p /content/drive/MyDrive

Executing: git clone --depth 1 https://gitlab.com/luigigenovese/bigdft-school
Error Occurred:  
 fatal: destination path 'bigdft-school' already exists and is not an empty directory.

Executing: mkdir -p /content/drive/MyDrive/bigdft-school



# Closing the environment

There are also commodity functions which unmount your google drive (to keep it in sync with the data that have been copied in it) and to erase every file that has been put in by this school project.

In [ ]:
#this command can be executed to clean completely the google drive (don't do this now, but ad the end of the school!)
# install.purge_drive()


Executing: rm -rf /content/drive/MyDrive/bigdft-school


In [ ]:
#we flush the google drive at the end of the process (this can take a long time. You can install locally the client if it takes too long)
install.close_drive()